### PR TITLE
Feature/bubble detector fix

### PR DESCRIPTION
**Pull request by tadzjibov**
Fixed bugs and performance problems with loading TB328

### CODE REVIEW

File: src/main/java/io/github/programminglife2016/pl1_2016/Launcher.java [hunk 1]
```diff
@@ -34,23 +33,36 @@ private Launcher() {
...
+        InputStream positions = Launcher.class.getResourceAsStream(
+                String.format("/genomes/%s.positions", dataset));
+        InputStream metadata = Launcher.class.getResourceAsStream("/genomes/metadata.csv");
+        SegmentParser segmentParser = new SegmentParser(positions, metadata);
+        NodeCollection nodeCollection = segmentParser.parse(is);
+
+        SetupDatabase sd = new SetupDatabase();
+        sd.setup(nodeCollection);
+//        Server server = new RestServer(queryStrategy);
+//        server.startServer();
```
Comment (smesseim) [inline src/main/java/io/github/programminglife2016/pl1_2016/Launcher.java]: Shouldn't the server be started?

File: src/main/java/io/github/programminglife2016/pl1_2016/collapser/BubbleDetector.java [hunk 2]
```diff
@@ -22,127 +23,142 @@
...
-                if (bubble.getStartNode().getLinks().size() == 1
-                        || bubble.getStartNode() == bubble.getEndNode()) {
-                    continue;
-                }
-                for (Node node : bubble.getStartNode().getLinks()) {
-                    List<Node> newBubbles = findLevelBubbles(node, bubble.getEndNode());
-                    currLevelList.addAll(newBubbles);
-                }
+        while (true) {
+//            conductedSearches = new HashSet<>();
```
Comment (smesseim) [inline src/main/java/io/github/programminglife2016/pl1_2016/collapser/BubbleDetector.java]: Remove pls

File: src/main/java/io/github/programminglife2016/pl1_2016/collapser/BubbleDetector.java [hunk 3]
```diff
@@ -22,127 +23,142 @@
...
-            for (Node bubble : levelBubbles.get(reachedLevel - 1)) {
-                if (bubble.getStartNode().getLinks().size() == 1
-                        || bubble.getStartNode() == bubble.getEndNode()) {
-                    continue;
-                }
-                for (Node node : bubble.getStartNode().getLinks()) {
-                    List<Node> newBubbles = findLevelBubbles(node, bubble.getEndNode());
-                    currLevelList.addAll(newBubbles);
-                }
+        while (true) {
```
Comment (smesseim) [inline src/main/java/io/github/programminglife2016/pl1_2016/collapser/BubbleDetector.java]: Can be rewritten to `while (!currLevelList.isEmpty())`. More readable that way.

File: src/main/java/io/github/programminglife2016/pl1_2016/collapser/BubbleDetector.java [hunk 4]
```diff
@@ -22,127 +23,142 @@
...
             reachedLevel++;
         }
-        if (levelBubbles.size() > 1) {
-            levelBubbles.remove(reachedLevel - 1);
-        }
-        this.bubbleBoundaries = new ArrayList<>(levelBubbles.get(1));
-        for (int i = 2; i < levelBubbles.size() + 1; i++) {
-            this.bubbleBoundaries.addAll(levelBubbles.get(i));
-        }
+        levelBubbles.values().stream().forEach(this.bubbleBoundaries::addAll);
```
Comment (smesseim) [inline src/main/java/io/github/programminglife2016/pl1_2016/collapser/BubbleDetector.java]: No need to convert the collection to a stream.

File: src/main/java/io/github/programminglife2016/pl1_2016/collapser/BubbleDetector.java [hunk 5]
```diff
@@ -175,32 +191,52 @@ private void handleDetectedBubble(Node starting, Node value, List<Node> levelCol
...
 
     /**
-     * Return the boundaries within which a bubble is detected. It filters
-     * out any dublicate or redundant bubbles that were found.
-     * @return List of bubbles that were detected.
+     * Get bubbles with unique pair of start and end node.
+     * @return list of unique bubbles
      */
-    final List<Node> getBubbleBoundaries() {
+    List<Node> getBubbleBoundaries() {
```
Comment (smesseim) [inline src/main/java/io/github/programminglife2016/pl1_2016/collapser/BubbleDetector.java]: Make it private.

File: src/main/java/io/github/programminglife2016/pl1_2016/collapser/BubbleDetector.java [hunk 6]
```diff
@@ -175,32 +191,52 @@ private void handleDetectedBubble(Node starting, Node value, List<Node> levelCol
...
+//    }
+//
+//    private void dfs(Node node) {
+//        for (Node link : node.getLinks()) {
+//            if (!visited[link.getId()]) {
+//                visited[link.getId()] = true;
+//                dfs(node);
+//            }
+//        }
+//    }
```
Comment (smesseim) [inline src/main/java/io/github/programminglife2016/pl1_2016/collapser/BubbleDetector.java]: Remove this pls.

File: src/main/java/io/github/programminglife2016/pl1_2016/collapser/BubbleLinker.java [hunk 7]
```diff
@@ -94,40 +141,36 @@ private void setLowerLevels(Node bubble) {
...
                     .filter(x -> x.getZoomLevel() == currLevel)
                     .collect(Collectors.toList());
-            int l = level.size();
-            for (int i = 0; i < l; i++) {
-                addLinkToBubble(level.get(i));
-            }
+//            int l = level.size();
+//            for (int i = 0; i < l; i++) {
+//                addLinkToBubble(level.get(i));
+//            }
```
Comment (smesseim) [inline src/main/java/io/github/programminglife2016/pl1_2016/collapser/BubbleLinker.java]: Remove

File: src/main/java/io/github/programminglife2016/pl1_2016/collapser/BubbleLinker.java [hunk 8]
```diff
@@ -154,12 +197,14 @@ private void addLinkToBubble(Node bubble) {
...
-                    .findFirst()
-                    .get());
+        if (prospectiveLink.getZoomLevel() > level
+            && quickReference.containsKey(String.valueOf(cId))) {
+            prospectiveLink = getHighestBubble(level, quickReference.get(String.valueOf(cId)));
+//                    bubbles
+//                    .parallelStream()
+//                    .filter(x -> x.getId() == cId)
+//                    .findFirst()
+//                    .get());
```
Comment (smesseim) [inline src/main/java/io/github/programminglife2016/pl1_2016/collapser/BubbleLinker.java]: Remove

File: src/main/java/io/github/programminglife2016/pl1_2016/collapser/BubbleLinker.java [hunk 9]
```diff
@@ -177,73 +222,110 @@ private boolean isEndNode(Node node) {
...
-        if (exist.isPresent()) {
-            return (Bubble) exist.get();
+    private synchronized Bubble initNewBubble(Node node, int level) {
+//            Optional<Node> exist = bubbles.parallelStream()
+//                    .filter(x -> x.getStartNode().getId() == node.getStartNode().getId()
+//                            && x.getEndNode().getId() == node.getEndNode().getId()
+//                            && x.getZoomLevel() == level).findFirst();
+//            if (exist.isPresent()) {
+//                return (Bubble) exist.get();
+//            }
```
Comment (smesseim) [inline src/main/java/io/github/programminglife2016/pl1_2016/collapser/BubbleLinker.java]: Remove

File: src/main/java/io/github/programminglife2016/pl1_2016/collapser/SimplePair.java [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+/**
+ * Class to store tuples.
+ * @param <T> first element of the pair.
+ * @param <E> second element of the pair.
+ * @author Ravi Autar
+ */
+public class SimplePair<T, E> {
+
+    private T t;
+    private E e;
```
Comment (smesseim) [inline src/main/java/io/github/programminglife2016/pl1_2016/collapser/SimplePair.java]: There are no accessors for these fields. What is this class used for?

File: src/main/java/io/github/programminglife2016/pl1_2016/database/FetchDatabase.java [hunk 11]
```diff
@@ -62,12 +62,16 @@ public void connect() {
...
+        String query = String.format("SELECT DISTINCT %s.* "
+                + "FROM %s, (SELECT DISTINCT n1.id AS from, n2.id AS to FROM "
+                + "" + "" + "%s AS n1 JOIN %s ON n1.id = %s.from_id "
+                + "JOIN %s AS n2 ON n2.id = %s.to_id WHERE %s"
+                + ".threshold =" + " %d AND" + " ((n1.x >= %d AND n1.x <= %d) "
+                + "OR (n2.x >= %d AND n2.x <= %d)))sub "
+                + "WHERE sub.from = %s" + ".id OR sub.to =" + " %s.id ORDER BY %s.id",
+                NODES_TABLE, NODES_TABLE, NODES_TABLE, LINK_TABLE, LINK_TABLE, NODES_TABLE,
+                LINK_TABLE, LINK_TABLE, threshold, x1, x2, x1, x2, NODES_TABLE,
+                NODES_TABLE, NODES_TABLE);
```
Comment (smesseim) [inline src/main/java/io/github/programminglife2016/pl1_2016/database/FetchDatabase.java]: Can you rewrite this to a prepared statement?

File: src/main/java/io/github/programminglife2016/pl1_2016/database/FetchDatabase.java [hunk 12]
```diff
@@ -150,10 +156,15 @@ private JSONArray fetchMetadata(String genome) {
...
-                + "<= %d) OR (n1.x <= %d AND n2.x >= %d))", NODES_TABLE, LINK_TABLE, LINK_TABLE, NODES_TABLE, LINK_TABLE, LINK_TABLE, threshold, x1,
+        String query = String.format("SELECT DISTINCT n1.id as "
+                + "from, n1.x AS x1, n1.y AS y1, n2.id as to, n2.x AS x2,"
+                + "" + "" + " n2.y AS y2 "
+                + "FROM %s AS n1 JOIN %s ON n1.id = %s.from_id JOIN %s AS n2 ON n2.id = %s"
+                + ".to_id " + "WHERE %s" + ".threshold = %d "
+                + "AND ((n1.x >= %d AND n1.x <= %d) OR (n2.x >= %d AND n2.x "
+                + "<= %d) OR (n1.x <= %d AND n2.x >= %d))",
+                NODES_TABLE, LINK_TABLE, LINK_TABLE, NODES_TABLE,
+                LINK_TABLE, LINK_TABLE, threshold, x1,
```
Comment (smesseim) [inline src/main/java/io/github/programminglife2016/pl1_2016/database/FetchDatabase.java]: Here too.

File: src/main/java/io/github/programminglife2016/pl1_2016/parser/nodes/SegmentParser.java [hunk 13]
```diff
@@ -51,6 +51,11 @@ public SegmentParser(InputStream positions, InputStream metadata) {
         lastIndices = new HashMap<>();
         SpecimenParser specimenParser = new SpecimenParser();
         this.specimens = specimenParser.parse(metadata);
+        Subject ref = new Specimen();
+        ref.setNameId(REFERENCE);
+        if (specimens != null) {
+            specimens.put(REFERENCE, ref);
+        }
```
Comment (smesseim) [inline src/main/java/io/github/programminglife2016/pl1_2016/parser/nodes/SegmentParser.java]: This part is not necessary any more: the missing genomes are inserted while parsing the segments.

File: src/main/java/io/github/programminglife2016/pl1_2016/parser/nodes/SegmentParser.java [hunk 14]
```diff
@@ -153,14 +158,26 @@ private void parseSegmentLine(String[] data) {
             nodeCollection.get(id).setData(seq);
             nodeCollection.get(id).setColumn(column);
         }
+
+        Set<Subject> genomes;
         if (specimens != null) {
```
Comment (smesseim) [inline src/main/java/io/github/programminglife2016/pl1_2016/parser/nodes/SegmentParser.java]: WTF is this.

File: src/main/resources/genomes/converter.py [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+#write .dot file
+w = open(OUTPUT + ".dot", "w")
+w.write("digraph G {\n\trankdir=LR;\n\tnodesep=1.2;\n\tnode[shape=circle];\n")
+for line in dot: w.write("\t" + line + "\n")
+w.write("}")
+w.close()
+
+#generate output from .dot file
+for fmt, ext in FORMATS:
+    os.system("dot -T{} {} -o {}".format(fmt, OUTPUT + ".dot", OUTPUT + ".{}".format(ext)))
```
Comment (smesseim) [inline src/main/resources/genomes/converter.py]: There is no need to include python scripts: they are only used once.

File: src/main/resources/genomes/gfaconverter.py [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+f = open(filename, "r").readlines()
+data = map(lambda line: re.split("\\s+", line), f)
+lines = filter(lambda line: line[0] == "L", data)
+gfa_lines = map(lambda x: x[1] + " -> " + x[3] + ";", lines)
+w = open(OUTPUT + ".dot", "w")
+w.write("digraph G {\n\trankdir=LR;\n\tnodesep=1.2;\n\tnode[shape=circle];\n")
+for line in gfa_lines: w.write("\t" + line + "\n")
+w.write("}")
+w.close()
+os.system("dot -T{} {} -o {}".format(FORMAT, OUTPUT + ".dot", OUTPUT + ".{}".format(FORMAT)))
```
Comment (smesseim) [inline src/main/resources/genomes/gfaconverter.py]: There is no need to include python scripts: they are only used once.

**Comment by smesseim** [comment]
Remove the feature files and the test files in resources.

**Comment by RaviShivam** [comment]
Well done, LGTM

**Pull request merged**